src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py

```diff
@@ -0,0 +1,153 @@
+from copy import deepcopy
+from typing import Union
+
+from omegaconf import DictConfig
+import torch
+
+from super_gradients import is_distributed
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.training import models
+from torch.distributed import barrier
+
+logger = get_logger(__name__)
+
+
+class PreLaunchCallback:
+    """
+    PreLaunchCallback
+
+    Base class for callbacks to be triggered, manipulating the config (cfg) prior to launching training,
+     when calling Trainer.train_from_config(cfg).
+
+    """
+
+    def __call__(self, cfg: Union[dict, DictConfig]) -> Union[dict, DictConfig]:
+        raise NotImplementedError
+
+
+class AutoTrainBatchSizeSelectionCallback(PreLaunchCallback):
+    """
+    AutoTrainBatchSizeSelectionCallback
+
+    Modifies cfg.dataset_params.train_dataloader_params.batch_size by searching for the maximal batch size that fits
+     gpu memory. Works out of the box for DDP.
+
+    The search is done by running a few forward passes for increasing batch sizes, until CUDA OUT OF MEMORY is raised:
+
+        For batch_size in range(min_batch_size:max_batch_size:size_step):
+            if batch_size raises CUDA OUT OF MEMORY ERROR:
+                return batch_size-size_step
+        return batch_size
+
+    Example usage: Inside the main recipe .YAML file (for example super_gradients/recipes/cifar10_resnet.yaml),
+     add the following:
+
+    pre_launch_callbacks_list:
```

Comment:
That's great addition to SG, but I try to think of the use cases and I find that adding the `pre_launch_callbacks_list` to config is slightly not I as a user would like to do.
Let me elaborate what I mean.
Suppose I have a recipe tailored for 8xA100. Batch size is hard-coded in YAML, num gpus as well.
I have 4x3090, that has less GPU memory on board.

So I may override this via command-line args:

`python train.py --config-name=my_recipe batch_size=16 num_gpus=4` (Let's assume we had 64 for A100)

This is our baseline. Whatever we're about to propose as a replacement should be AT LEAST as simple and as useful what user can do via cmdline.

Ideally, if I as a user can write this:

`python train.py --config-name=my_recipe batch_size=auto num_gpus=all`

And under the hood, these `auto`/`all` should trigger some PreLaunchCallback to infer num GPUS and batch size. 

Additionally, it is just impossible to add an item to `pre_launch_callbacks_list` list via command-line interface. So we are forcing user to add these lines to config file which may not be the in his intent. 

Certainly, for recipes that are meant to be executed on various setups it make sense to include `AutoTrainBatchSizeSelectionCallback` in the file. I can see three use-cases where we can use this feature:
* AutoNAC / CI - For this use-case the proposed approach is totally ok.
* Colab Notebooks - There guys certainly not going to modify recipes. And at most they can type `Trainer(..., batch_size="auto")`
* Regular users trying to fit recipe to their HW. I think they can do both command-line override and config modification, however command-line override is def. faster.

Reply:
Thanks for the feedback (:

I completely agree that setting this option via command line is a good idea.
However, notice that there are multiple settings in which we can run the batch size selector callback, and they are quite sensitive (in the sense that setting defaults here might cause you to run forever until finding the right batch size). I am not confident this is the desired behaviour. Perhaps it's a good idea to ask around and see what answers we get.

Addressing the usages:
- We both agree for external use (i.e when someone has structured their own recipe) this is already useful.
- Now regarding Colab notebooks:
note that for for the callback to work, we are launching Trainer.train_from_config(), this is because we must access to dataset_params and modify them accordingly, as well as the learning rate which is in training params. My point here is that for coded train launch- we cant really offer auto batch size selection (i.e by just calling Trainer.train()) because Trainer.train() expects instantiated dataloaders. Re-instantation of (possibly external) dataloaders with a different batch size can get messy, especially when using samplers. With that being said, its probably possible, so if you have a suggestion how to do it I'm all ears (: .
I do have the feeling it would add much more complexity for a case that might not bee so common.
Also note that users can still use Trainer.train_from_config assuming they have their own config yaml files in the notebook file system somewhere.

-Third point- regular users using our recipes- this point I actually find the most interesting. We surely can add a shortcut for it. My concern is that since we don't have a "recipe base" at the moment, this might lead users to have errors if they did not inherit from our base recipe config (since they won't have "batch_size" arg in the main config file).

Lets further discuss these points.